File: BaekJoon/2573.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <utility>

using namespace std;

int n, m;
int mr[4] = {0, 0, 1, -1};
int mc[4] = {1, -1, 0, 0};

// 0: 모든 빙산이 녹아버림, else: 현재 빙산의 개수
int countIceberg(vector<vector<int>> &map)
{
  vector<vector<bool>> visited(n, vector<bool>(m, false));
  queue<pair<int, int>> taskQueue;
  int ret = 0;

  for (int i = 0; i < n; i++)
  {
    for (int j = 0; j < m; j++)
    {
      if (map[i][j] == 0 || visited[i][j])
      {
        continue;
      }
      taskQueue.push({i, j});
      visited[i][j] = true;
      ret++;
      while (!taskQueue.empty())
      {
        int r = taskQueue.front().first;
        int c = taskQueue.front().second;
        taskQueue.pop();
        for (int k = 0; k < 4; k++)
        {
          int nextR = r + mr[k];
          int nextC = c + mc[k];
          if (nextR < 0 || nextC < 0 || nextR >= n || nextC >= m ||
              visited[nextR][nextC] || map[nextR][nextC] == 0)
          {
            continue;
          }
          visited[nextR][nextC] = true;
          taskQueue.push({nextR, nextC});
        }
      }
    }
  }
  return ret;
}

void meltIceberg(vector<vector<int>> &map)
{
  vector<vector<int>> meltingScale(n, vector<int>(m, 0));

  for (int r = 0; r < n; r++)
  {
    for (int c = 0; c < m; c++)
    {
      if (map[r][c] == 0)
      {
        continue;
      }
      for (int k = 0; k < 4; k++)
      {
        int nextR = r + mr[k];
        int nextC = c + mc[k];
        if (nextR < 0 || nextC < 0 || nextR >= n || nextC >= m || map[nextR][nextC] != 0)
        {
          continue;
        }
        meltingScale[r][c]++;
      }
    }
  }
  for (int r = 0; r < n; r++)
  {
    for (int c = 0; c < m; c++)
    {
      map[r][c] = max(map[r][c] - meltingScale[r][c], 0);
    }
  }
}

void printMap(vector<vector<int>> &map)
{
  for (int r = 0; r < n; r++)
  {
    for (int c = 0; c < m; c++)
    {
      cout << map[r][c] << ' ';
    }
    cout << '\n';
  }
}

int main()
{
  // freopen("./input.txt", "r", stdin);
  ios_base::sync_with_stdio(0);
  cin.tie(0);
  cin >> n >> m;
  vector<vector<int>> map(n, vector<int>(m));

  for (int r = 0; r < n; r++)
  {
    for (int c = 0; c < m; c++)
    {
      cin >> map[r][c];
    }
  }
  int elapsedYears = 0;
  while (true)
  {
    elapsedYears++;
    meltIceberg(map);
    int numberIcebergs = countIceberg(map);
    if (numberIcebergs == 0)
    {
      cout << 0;
      break;
    }
    else if (numberIcebergs != 1)
    {
      cout << elapsedYears;
      break;
    }
  }

  return 0;
}

File: BaekJoon/2573.py
import sys

# sys.stdin = open("input.txt", "r")

n, m = map(int, input().split())

ice = [list(map(int, input().split())) for _ in range(n)]

dx = [0, 0, 1, -1]
dy = [1, -1, 0, 0]


def check():
    visited = [[False] * m for _ in range(n)]
    cnt = 0
    for i in range(n):
        for j in range(m):
            if ice[i][j] > 0 and not visited[i][j]:
                cnt += 1
                stack = [(i, j)]
                visited[i][j] = True
                while stack:
                    x, y = stack.pop()
                    for k in range(4):
                        nx, ny = x + dx[k], y + dy[k]
                        if (
                            0 <= nx < n
                            and 0 <= ny < m
                            and ice[nx][ny] > 0
                            and not visited[nx][ny]
                        ):
                            stack.append((nx, ny))
                            visited[nx][ny] = True
    return cnt


def melt():
    melt_level = [[0] * m for _ in range(n)]
    melt_cnt = 0
    for i in range(n):
        for j in range(m):
            if ice[i][j] == 0:
                continue

            melt_cnt += 1

            for k in range(4):
                ni, nj = i + dx[k], j + dy[k]
                if 0 <= ni < n and 0 <= nj < m and ice[ni][nj] == 0:
                    melt_level[i][j] += 1

    for i in range(n):
        for j in range(m):
            ice[i][j] = max(0, ice[i][j] - melt_level[i][j])

    return melt_cnt


year = 0
while True:
    if melt() == 0:
        print(0)
        break

    year += 1

    if check() >= 2:
        print(year)
        break
